Plot recipe for VectorField

schillic opened the pull request:
This is open for discussion.

Usage:
```julia
julia> f(x) = [2*x[2], -x[1]]
julia> V = MathematicalSystems.VectorField(f)
julia> plot(V)  # uses a default grid of 20x20 points for the range [-3, 3] x [-3, 3]
```
![1](https://user-images.githubusercontent.com/9656686/100524361-d6f3a500-31b7-11eb-93e8-f53e45af05e2.png)
```julia
julia> plot(V, x=range(-3, stop=3, length=11), y=range(-3, stop=3, length=11))  # pass a custom grid
```
![2](https://user-images.githubusercontent.com/9656686/100524364-d955ff00-31b7-11eb-9867-ae981594a2d1.png)

EDIT: For reference, this is the result of the default quiver plot (note that here the function `f` needs to take two arguments and in my approach it needs to take a vector):

```julia
f(x, y) = [2*x, -y]
x = range(-3, stop=3, length=21)
y = range(-3, stop=3, length=21)
x_full = [xi for xi in x for yj in y]
y_full = [yj for xi in x for yj in y]
quiver(x_full, y_full, quiver=f)
```
![3](https://user-images.githubusercontent.com/9656686/100524549-664d8800-31b9-11eb-9afb-fa15fe0f3b04.png)

mforets commented on src/vector_field.jl:
```diff
@@ -163,3 +163,71 @@ function VectorField(sys::AbstractContinuousSystem)
+@recipe function plot(V::VectorField;
+                      x=range(-3, stop=3, length=21),
+                      y=range(-3, stop=3, length=21))
+    seriestype := :quiver
```
check the the vector field is 2D and give an error otherwise?

schillic replied:
Actually, you can also plot a 2D projection. For that you could be able to pass the two dimensions that you want to plot, probably defaulting to `1` and `2`.

schillic replied:
I did that now. You can do `f(x) = [2*x[2], -x[1], 7]` in my example above and `plot(V)` will give the same result. You can do `plot(V, dims=[1, 3])` to change the dimensions.

mforets commented on src/vector_field.jl:
```diff
@@ -163,3 +163,71 @@ function VectorField(sys::AbstractContinuousSystem)
+
+### plot recipe
+@recipe function plot(V::VectorField;
+                      x=range(-3, stop=3, length=21),
```
i think it can be useful if we also allow the kwargs `xmin`, `xmax`, `ymin`, `ymax`? or can we just take `xlims` and `ylims` which are already valid Plots kwargs?

schillic replied:
So define a default number of grid points you mean (the `length` argument in my `range` above)? That is probably reasonable because when I experimented with different numbers, there is only a limited reasonable range of 10-30 grid points in each dimension.

schillic replied:
I combined and renamed the arguments to `grid_points`, which should be more understandable. (`grid` is already reserved by `Plots`.)

mforets commented on src/vector_field.jl:
```diff
@@ -163,3 +163,71 @@ function VectorField(sys::AbstractContinuousSystem)
+    N_VAL = typeof(V([x1, y1])[1])
+    m = length(x) * length(y)
+
+    sx = Vector{N_POS}(undef, m)
```
does the `s` or the `t` represents widths along each direction of the arrows? can you please add a short description of what the function is doing?

schillic replied:
I think when I wrote the code I had `s` and `t` for source and target coordinates of the arrow in mind, but I may be wrong.

schillic replied:
No, I was wrong. I renamed the variables and added a small documentation (just as a comment).

mforets reviewed (commented):
Nice!!

schillic commented:
I will probably not have time to work on this in the near future. Feel free to modify.

schillic commented:
Side note: `VectorField` is not exported. I do not remember whether that was on purpose or an oversight.

mforets commented:
> I will probably not have time to work on this in the near future. Feel free to modify.

:+1:  me too and since the others replied with thumbs up and the PR is functional, I suggest to proceed merging it and addressing a documentation PR separately. (Please note that the build is unrelated and only affects Julia v1.0).

PS: The new function may interest @dfcaporale

schillic commented:
I fixed the build issue and addressed your comments.
I do not mind keeping this open a bit until we agree on the API. I mainly wanted to share the code and not forget about it.

mforets commented:
Merging.. I don't mind to follow up with API changes / docs updates in subsequent PRs.
